### PR TITLE
Puppeth: add option to specify geth build repo for node deployment

### PR DESCRIPTION
**Pull request by lightclient**
Currently, only the latest version of the `ethereum:go-client` image is used to deploy nodes. This PR adds support for building images from custom geth repositories.

### CODE REVIEW

File: cmd/puppeth/wizard_node.go [hunk 1]
```diff
@@ -77,6 +77,35 @@ func (w *wizard) deployNode(boot bool) {
...
+	case choice == "2":
+		fmt.Println()
+		fmt.Println("Where's the git repository? (http/https url)")
+		url := w.readURL()
+
+		switch url.Scheme {
+		case "http", "https":
+			infos.gitRepo = url.String()
+			fmt.Println()
+			fmt.Println("What branch? (default = repository default)")
```
Comment (karalabe) [inline cmd/puppeth/wizard_node.go:98]: You specify that `empty` will be the default branch, but your clone command always specified the branch `--branch {{.GitBranch}}`. As such, if none is given, the clone will fail

```
$ git clone https://github.com/ethereum/go-ethereum --branch --single-branch
Cloning into 'go-ethereum'...
fatal: Remote branch --single-branch not found in upstream origin
```

File: cmd/puppeth/module_node.go [hunk 2]
```diff
@@ -32,8 +32,24 @@ import (
 
 // nodeDockerfile is the Dockerfile required to run an Ethereum node.
 var nodeDockerfile = `
-FROM ethereum/client-go:latest
+{{if .Git}}
+	FROM golang:1.16-alpine as builder
 
+	RUN apk add --no-cache make gcc musl-dev linux-headers git
+
+	RUN git clone {{.GitRepo}} --branch {{.GitBranch}} --single-branch /go-ethereum
```
Comment (karalabe) [inline cmd/puppeth/module_node.go:40]: It probably also makes sense to add `--depth=1` to not download the entire commit history, just the tip.

Comment (lightclient) [reply cmd/puppeth/module_node.go:40]: Good idea.

**Comment by lightclient** [comment]
I've been using this for a couple weeks now, and I'm not convinced it's the best solution. It does allow puppeth to deploy custom versions, which AFAICT is not currently possible, but if you've modified the genesis file you must be using a version of puppeth compiled with those changes. The reason being, if you a preconfigured genesis, puppeth wants to parse it into a genesis object. When you deploy a node, puppeth serializes the genesis object back into json. But if you're using a version of puppeth that doesn't support the new fork block that you've added, it will not decode you fork block into a genesis object (since no matching field exists) and therefore when it prepares to launch a node, it doesn't write the same genIMO, it seems unnecessary tesis as it read.

It seems like there are a few solutions:
* show users a warning when they try to deploy a geth version on a different commit than puppeth - this is currently what's implemented and it's not ideal
* add the option to not decode the genesis file and just copy it to the remote server when needed for geth instances
* store raw genesis by default and copy to server byte-for-byte when deploying a node - this makes exporting other client genesis files sort of weird since they will not include the new fork block
* other?

**Comment by lightclient** [comment]
Closing for now - maybe come up with a better solution another time.

**Pull request closed without merging**